includes/AMP/Canonical_Sanitizer.php

```diff
@@ -54,28 +54,43 @@ class Canonical_Sanitizer extends AMP_Base_Sanitizer {
 	 * @return void
 	 */
 	public function sanitize() {
-		$links         = [];
+		/**
+		 * The rel=canonical DOM element.
+		 *
+		 * @var DOMElement The element.
+		 */
+		$rel_canonical = null;
 		$link_elements = $this->dom->head->getElementsByTagName( Tag::LINK );
 
+		// This fallback to get_permalink() ensures that there's a canonical link
+		// even when previewing drafts.
+		$canonical_url = wp_get_canonical_url();
+		if ( ! $canonical_url ) {
+			$canonical_url = get_permalink();
+		}
+
 		/**
 		 * Link element.
 		 *
 		 * @var DOMElement $link
 		 */
 		foreach ( $link_elements as $link ) {
-			if ( $link->hasAttribute( Attribute::REL ) ) {
-				$links[ $link->getAttribute( Attribute::REL ) ][] = $link;
+			if ( $link->getAttribute( Attribute::REL ) === Attribute::REL_CANONICAL ) {
+				if ( ! empty( $link->getAttribute( Attribute::HREF ) ) ) {
+					return;
+				}
+
+				$link->setAttribute( Attribute::HREF, (string) $canonical_url );
```

Comment:
Can we check if it is empty before setting it? What if it is false?

Reply:
According to the docs, `get_permalink()` only returns `false` if the post does not exist, which at this point should never be the case.

Worst case the `href` attribute was empty before and now is still empty, which would be the same situation as before this PR, on `main`.